filters/imagefilter.go

```diff
@@ -43,65 +41,74 @@ func init() {
 
 type ImageFilter interface {
 	CreateOptions(image *bimg.Image) (*bimg.Options, error)
+	CanBeMerged(other *bimg.Options, self *bimg.Options) bool
+	Merge(other *bimg.Options, self *bimg.Options) *bimg.Options
 }
 
 func HandleImageResponse(ctx filters.FilterContext, f ImageFilter) {
-	rsp := ctx.Response()
-
-	defer rsp.Body.Close()
-
-	rsp.Header.Del("Content-Length")
-
-	r, err := handleImageTransform(rsp.Body, f)
+	image := ctx.StateBag()[SkropImage].(*bimg.Image)
 
+	opt, err := f.CreateOptions(image)
 	if err != nil {
-		log.Error("failed to process image ", err.Error())
+		log.Error("Failed to create options ", err.Error())
 		ctx.Serve(&http.Response{
-			StatusCode:http.StatusInternalServerError,
-			Body: ioutil.NopCloser(bytes.NewBufferString(err.Error())),
+			StatusCode: http.StatusInternalServerError,
+			Body:       ioutil.NopCloser(bytes.NewBufferString(err.Error())),
 		})
 		return
 	}
 
-	rsp.Body = ioutil.NopCloser(r)
-}
+	opts := ctx.StateBag()[SkropOptions].(*bimg.Options)
 
-func handleImageTransform(r io.Reader, f ImageFilter) (io.Reader, error) {
-	var err error
-	imageBytes, err := ioutil.ReadAll(r)
+	if f.CanBeMerged(opts, opt) {
+		ctx.StateBag()[SkropOptions] = f.Merge(opts, opt)
+		log.Debug("Filter ", f, " merged in ", ctx.StateBag()[SkropOptions])
+		return
+	}
 
+	//transform the image
+	buf, err := transformImage(image, opts)
 	if err != nil {
-		return nil, err
+		log.Error("Failed to process image ", err.Error())
+		ctx.Serve(&http.Response{
+			StatusCode: http.StatusInternalServerError,
+			Body:       ioutil.NopCloser(bytes.NewBufferString(err.Error())),
+		})
+		return
 	}
 
-	imageBytesLength := len(imageBytes)
+	// set opt in the stateBag
+	ctx.StateBag()[SkropImage] = bimg.NewImage(buf)
+	ctx.StateBag()[SkropOptions] = opt
 
-	log.Debug("Image bytes length: ", imageBytesLength)
-
-	if imageBytesLength == 0 {
-		return nil, errors.New("original image is empty. nothing to process")
-	}
-
-	originalImage := bimg.NewImage(imageBytes)
+}
 
-	options, err := f.CreateOptions(originalImage)
+func FinalizeResponse(ctx filters.FilterContext) {
+	image := ctx.StateBag()[SkropImage].(*bimg.Image)
+	opts := ctx.StateBag()[SkropOptions].(*bimg.Options)
 
+	buf, err := transformImage(image, opts)
 	if err != nil {
-		return nil, err
+		log.Error("Failed to process image ", err.Error())
+		ctx.Serve(&http.Response{
+			StatusCode: http.StatusInternalServerError,
+			Body:       ioutil.NopCloser(bytes.NewBufferString(err.Error())),
+		})
+		return
 	}
 
-	transBytes, err := transformImage(originalImage, options)
+	rsp := ctx.Response()
 
-	if err != nil {
-		return nil, err
-	}
+	defer rsp.Body.Close()
```

Comment:
defer rsp.Body.Close should be called before the first return, to make sure that it's always closed. Not closing it may cause the backend connection hanging.